devito/ir/support/basic.py

```diff
@@ -292,50 +285,48 @@ def lex_le(self, other):
     def lex_lt(self, other):
         return self.timestamp < other.timestamp
 
-    def distance(self, other, findex=None):
+    def distance(self, other):
         """
         Compute the distance from ``self`` to ``other``.
 
         Parameters
         ----------
         other : TimedAccess
-            The TimedAccess from which the distance is computed.
-        findex : Dimension, optional
-            If supplied, compute the distance only up to and including ``findex``.
+            The TimedAccess w.r.t. which the distance is computed.
         """
-        assert isinstance(other, TimedAccess)
-
-        if not self.rank:
-            return Vector()
-
-        # Compute distance up to `limit`, ignoring `directions` for the moment
-        if findex is None:
-            findex = self.findices[-1]
-            limit = self.rank
-        else:
-            try:
-                limit = self._cached_findices_index[findex] + 1
-            except KeyError:
-                raise TypeError("Cannot compute distance as `findex` not in `findices`")
-
         ret = []
-        for n, (i, o) in enumerate(zip(self[:limit], other)):
+        for sit, oit in zip(self.itintervals, other.itintervals):
+            n = len(ret)
+
             try:
-                iit = self.itintervals[n]
-                oit = other.itintervals[n]
+                sai = self.aindices[n]
+                oai = other.aindices[n]
             except IndexError:
-                # E.g., self=R<u,[t+1, ii_src_0+1, ii_src_1+2]>
-                #       itintervals=(time, p_src)
+                # E.g., `self=R<f,[x]>` and `self.itintervals=(x, i)`
                 break
 
-            if iit.direction is oit.direction and iit.interval == oit.interval:
-                if iit.direction is Backward:
-                    # Backward direction => flip the sign
-                    ret.append(o - i)
+            try:
+                if not (sit == oit and sai.root is oai.root):
+                    # E.g., `self=R<f,[x + 2]>` and `other=W<f,[i + 1]>`
+                    # E.g., `self=R<f,[x]>`, `other=W<f,[x + 1]>`,
+                    #       `self.itintervals=(x<0>,)` and `other.itintervals=(x<1>,)`
+                    ret.append(S.Infinity)
+                    break
+            except AttributeError:
+                # E.g., `self=R<f,[cy]>` and `self.itintervals=(y,)` => `sai=None`
+                pass
+
+            ai = sai
+            fi = self.findices[n]
+
+            if not ai or ai._defines & sit.dim._defines:
+                # E.g., `self=R<f,[t + 1, x]>`, `self.itintervals=(time, x)` and `ai=t`
+                if sit.direction is Backward:
+                    ret.append(other[n] - self[n])
                 else:
-                    ret.append(i - o)
-            else:
-                # Mismatching `itinterval` => Infinity
+                    ret.append(self[n] - other[n])
+            elif fi in sit.dim._defines:
```

Comment:
is that the only too possible cases? No default `else` for uncaught ones?

Reply:
no, if none of the two ifs are entered (the `if` and the `elif`), then we should just keep iterating